src/state_machine.rs

```diff
@@ -0,0 +1,1062 @@
+                         $crate::state_machine::Mutator::boxed(
+                             $crate::state_machine::MapMutator::new(
+                                 stringify!($name),
+                                 $crate::sugar::NamedArguments(names, values),
+                                 |$statevar: &mut $statety,
+                                  $crate::sugar::NamedArguments(
+                                      _, proptest_helper!(@_WRAPPAT ($($parm)*)))
+                                 | {
+                                     let ret: &mut $childty = $body;
+                                     Ok(ret)
+                                 },
+                                 mutation))))));
+
+        prop_sm_helper! {
+            @_MUNCH($ops, $invariants, $statety) $($rest)*
+        }
+    };
+
+    (@_MUNCH($ops:ident, $invariants:ident, $statety:ty)
+     $(#[$attr:ident = $attrv:expr])*
+     delegate ($child_machine:expr)
+     to $name:ident (&mut $statevar:ident, $($arg:tt)*)
+                     -> &mut $childty:ty $body:block
+     $($rest:tt)*) => {
+        let names = proptest_helper!(@_EXT _STR ($($arg)*));
+
+        #[allow(unused_mut)]
+        let mut attrs = $crate::state_machine::helper::OpAttrs::DEFAULT;
+        $(attrs.$attr = $attrv;)*
+
+        $ops.push(
+            (attrs.weight,
+             $crate::strategy::Strategy::boxed(
+                 $crate::strategy::Strategy::prop_map(
+                     ($child_machine.mutation_strategy(),
+                      proptest_helper!(@_EXT _STRAT ($($arg)*))),
+                     move |(mutation, values)|
+                         $crate::state_machine::Mutator::boxed(
+                             $crate::state_machine::MapMutator::new(
+                                 stringify!($name),
+                                 $crate::sugar::NamedArguments(names, values),
+                                 |$statevar: &mut $statety,
+                                  $crate::sugar::NamedArguments(
+                                      _, proptest_helper!(@_EXT _PAT ($($arg)*)))
+                                 | {
+                                     let ret: &mut $childty = $body;
+                                     Ok(ret)
+                                 },
+                                 mutation))))));
+
+        prop_sm_helper! {
+            @_MUNCH($ops, $invariants, $statety) $($rest)*
+        }
+    };
+}
+
+/// A state-machine-based test.
+///
+/// `StateMachine` values are simultaneously a description of mutations and
+/// invariants of that state, as well as a test that the state under test
+/// upholds its expected behaviours.
+///
+/// Instances of this struct are usually constructed by the
+/// `proptest_state_machine!` macro. Please see the documentation of that macro
+/// for examples of typical usage.
+#[must_use = "StateMachine does nothing unless used"]
+pub struct StateMachine<S : fmt::Debug> {
```

Comment:
A bound `S: 'static` is implied here because of WF (well-formedness) conditions on the fields...